### DISCUSSION TITLE
Probe output not working

### DISCUSSION THREAD

**Discussion by cooluser26** [epochpic/epoch#577] [Q&A] [answered]
I am putting probes at different points in EPOCH 1D simulation. However, with my recent input file, the output SDF file does not contain the probes output. I'm not sure why this is the case. Please help! The input file is below: 
```

begin:control
   nx = 4000

   x_min = 0
   x_max = 500e-6

   # Final time of simulation. 
   t_end = 2e-11

   stdout_frequency = 8000
end:control


begin:boundaries
   bc_x_min = periodic
   bc_x_max = periodic
end:boundaries


begin:constant
   dens_background = 1e+27
   #dens_beam = 1e+22
   yield = 8.4e+10
   sigmax = 2e-7
   mux = 0 
   dens_beam = 2*yield/(sqrt(pi) * sigmax)
   
   n_t = 100

   mass_ratio = 1836

   npart_background = 4*nx
   npart_beam = 10*nx
   
   drift_deuteron = 9.8*1e-20
   drift_triton = 1.1*1e-19
   drift_electron = 0
   drift_silicon = 0
   drift_oxygen = 0
   
   
   temp_ev_beam = 0
   temp_ev_background = 2.4*1e+3
   

end:constant


#######################
#### SPECIES BLOCK ####
#######################

begin:species
   # Rightwards traveling deuterons
   name = deuteron
   charge = +1
   mass = 2*mass_ratio
   
   temp_ev = temp_ev_beam
   drift_x = drift_deuteron
   number_density = dens_beam*gauss(x, mux, sigmax)   
   npart = npart_beam
end:species


begin:species
   # Rightwards traveling tritons
   name = triton
   charge = +1
   mass = 3*mass_ratio
   temp_ev = temp_ev_beam
   drift_x = drift_triton
   number_density = dens_beam*gauss(x, mux, sigmax) 
   npart = npart_beam
end:species

begin:species
   # background electrons
   name = background_electron
   charge = -1
   mass = 1
   temp_ev= temp_ev_background
   drift_x = drift_electron
   number_density = dens_background
   npart = npart_background
end:species

begin:species
   # background silicon/ablator ion. Assume v = 0 
   name = silicon
   charge = +14
   mass = 28*mass_ratio
   temp_ev= temp_ev_background
   drift_x = drift_silicon 
   number_density = dens_background*14 / (14+8)
   npart = npart_background
end:species

begin:species
   # background oxygen/ablator ion. Assume v = 0 
   name = oxygen
   charge = +8
   mass = 16*mass_ratio
   temp_ev= temp_ev_background
   drift_x = drift_oxygen
   number_density = dens_background*8 / (14+8)
   npart = npart_background
end:species

###########################################
####PROBE BLOCKS AT DIFFERENT DISTANCES####
###########################################

begin:probe
   name = triton_100_probe
   point = 100e-6
   normal = (1)

   ek_min = 0.0
   ek_max = -1.0
   include_species : triton
   dumpmask = always
end:probe

begin:probe
   name = triton_500_probe
   point = 500e-6
   normal = (1)  
   ek_min = 0.0
   ek_max = -1.0
   include_species : triton
   dumpmask = always
end:probe

begin:probe
   name = deuteron_100_probe
   point = 100e-6
   normal = (1)

   ek_min = 0.0
   ek_max = -1.0
   include_species : deuteron
   dumpmask = always
end:probe


begin:probe
   name = deuteron_500_probe
   point = 500e-6
   normal = (1)  

   ek_min = 0.0
   ek_max = -1.0
   include_species : deuteron
   dumpmask = always
end:probe


######################
### OUTPUT BLOCK  ####
######################
begin:output
   # Number of timesteps between output dumps
   dt_snapshot = t_end/n_t

   # Properties at particle positions
   particles = always
   
   vx= always          
   vy= always          
   vz= always 

   # Properties on grid
   grid = always
   ex= always
   ey = always
   ez = always

   bx= always
   by = always
   bz = always

   charge_density = always
   number_density = always + species
    
   particle_probes = always 
end:output

```

**Comment by Status-Mirror** [comment]
Hey @wnguyen1312,

I've run the simulation myself, and I do see probe outputs. For example, `0024.sdf` tracks 3672 macro-particles through `triton_100_probe`.

The probe outputs only exist if macro-particles have passed the probe since the previous output. For example, the 3672 macro-particles I mention have passed the probe in times between the write-time of `0023.sdf` and `0024.sdf` (4.6 ps and 4.8 ps). If no macro-particles pass a probe in this temporal window, then nothing is written to the SDF file.

Hope this helps,
Stuart

**Reply by cooluser26** [reply]
Hi Stuart,

Thank you for your explanation. I must have misunderstood how the probes work. 

I would like to use probes to measure the energy of particles as they traverse a particular plane during the simulation. The objective is to continuously track the energies throughout the entire duration of the simulation. My current understanding was that probes would only register readings if there's an interaction between time steps, which may not capture all the data I need. Would there be a work-around for this issue?

**EDIT**: I realized this could be done by setting it so we only have up to 0001.sdf file so all the entries are recorded there. Is this a valid way of doing it?

Thank you,

William.

**Reply by Status-Mirror** [reply, marked as answer]
Hey William @wnguyen1312,

In the previous method, all probe interactions are still recorded, they're just scattered about all the SDF files. You could still recover all the probe hits in the simulation if you scanned all the SDF files.

Yes, you can just dump a single SDF file at the end of the simulation, which would contain all the probe hits between the start and simulation end. The only disadvantage to this is that you lose temporal resolution - you don't know _when_ the particles pass the probes, which you would have with the multiple SDF method.

However, if the probe times are important, you can switch on the `-DPROBE_TIME` compiler flag in EPOCH. This forces probe particles to output an extra variable to tell you the exact time each particle passes the probe (which is better than the temporal window of multiple files). With a single-output, and `-DPROBE_TIME`, I think you should have full functionality.

Hope this helps,
Stuart

**Reply by cooluser26** [reply]
Using the `-DPROBE_TIME` compiler flag in EPOCH was exactly what I needed. Thanks a lot for your help, Stuart!
